### DISCUSSION TITLE
TypeError: 'int' object is not callable in PYG HeteroConv

### DISCUSSION THREAD

**Discussion by amalislam675** [pyg-team/pytorch_geometric#8153] [General]
I apply the HetroConv via this below mentioned code, but every time I find this error, 'int' object is not callable. Can you help me to fix this error?     
 Exact error message is :
`Traceback (most recent call last):
  File "/hetegeneousgraph/code/src/method/model1/main.py", line 248, in <module>
    train(model, optimizer, train_data)
  File "/hetegeneousgraph/code/src/method/model1/main.py", line 58, in train
    trn_loss = train_epoch()
  File "/hetegeneousgraph/code/src/method/model1/main.py", line 40, in train_epoch
    score = model(data.x_dict, data.edge_index_dict, data.edge_weight_dict)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/hetegeneousgraph/code/src/method/model1/model1.py", line 28, in forward
    h = self.conv1(x_dict, edge_index_dict, edge_weight_dict)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch_geometric/nn/conv/hetero_conv.py", line 136, in forward
    out = conv(x_dict[src], edge_index, *args, **kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch_geometric/nn/conv/gcn_conv.py", line 211, in forward
    edge_index, edge_weight, x.size(self.node_dim),
TypeError: 'int' object is not callable`                                    

The code which I used is this one, which is shown below:
`import torch
from torch_geometric.nn import GCNConv, HeteroConv

class Model(torch.nn.Module):
  def __init__(self, out_channels, hidden_channels=512):
    super().__init__()

    def edge_weight_fn(edge_index, node_features):
      # Compute the edge weights
      edge_weights = torch.ones(edge_index.size(1), device=edge_index.device)
      return edge_weights

    self.conv1 = HeteroConv({
        ('ntype1', 'etype1', 'ntype1'): GCNConv(8704, hidden_channels, edge_weight=edge_weight_fn),
        ('ntype2', 'etype2', 'ntype2'): GCNConv(8704, hidden_channels, edge_weight=edge_weight_fn),
        ('ntype3', 'etype3', 'ntype3'): GCNConv(8704, hidden_channels, edge_weight=edge_weight_fn)
      }, aggr='sum')
    self.conv2 = HeteroConv({
        ('ntype1', 'etype1', 'ntype1'): GCNConv(hidden_channels, hidden_channels, edge_weight=edge_weight_fn),
        ('ntype2', 'etype2', 'ntype2'): GCNConv(hidden_channels, hidden_channels, edge_weight=edge_weight_fn),
        ('ntype3', 'etype3', 'ntype3'): GCNConv(hidden_channels, hidden_channels, edge_weight=edge_weight_fn)
      }, aggr='sum')

    self.lin = Linear(hidden_channels*3, out_channels)

  def forward(self, x_dict, edge_index_dict, edge_weight_dict):

    h = self.conv1(x_dict, edge_index_dict, edge_weight_dict)
    h = {key: x.relu() for key, x in h.items()}
    h = self.conv2(h, edge_index_dict, edge_weight_dict)
    h = {key: x.relu() for key, x in h.items()}

    # Concatenate node embeddings from different node types
    concatenated_h = torch.cat([h[key] for key in h.keys()], dim=1)

    # Apply the linear layer to the concatenated embeddings
    out = self.lin(concatenated_h)

    return out`

My data Object is of this shape
`HeteroData(
  ntype1={ x=[8704, 8704] },
  ntype2={ x=[8704, 8704] },
  ntype3={ x=[8704, 8704] },
  (ntype1, etype1, ntype1)={
    edge_index=[2, 1511590],
    edge_weight=[1511590]
  },
  (ntype2, etype2, ntype2)={
    edge_index=[2, 1645254],
    edge_weight=[1645254]
  },
  (ntype3, etype3, ntype3)={ edge_index=[2, 322990] },
  (ntype3, etype3, ntype3)={ edge_weight=[322990] }
)
`
main function
`model =  Model(out_channels=5)
out = model(data.x_dict, data.edge_index_dict, data.edge_weight_dict)`

**Comment by rusty1s** [comment]
Mh, this looks correct to me (except for the `edge_weight` argument in the definition of `GCNConv`). Which version of PyG are you using? This example here runs fine for me:
```python
import torch

from torch_geometric.data import HeteroData
from torch_geometric.nn import GCNConv, HeteroConv

conv = HeteroConv(
    {
        ('ntype1', 'etype1', 'ntype1'): GCNConv(-1, 32),
        ('ntype2', 'etype2', 'ntype2'): GCNConv(-1, 32),
        ('ntype3', 'etype3', 'ntype3'): GCNConv(-1, 32),
    }, aggr='sum')

data = HeteroData()
data['ntype1'].x = torch.randn(10, 16)
data['ntype2'].x = torch.randn(10, 16)
data['ntype3'].x = torch.randn(10, 16)
data['ntype1', 'etype1', 'ntype1'].edge_index = torch.randint(0, 10, (2, 20))
data['ntype1', 'etype1', 'ntype1'].edge_weight = torch.rand(20)
data['ntype2', 'etype2', 'ntype2'].edge_index = torch.randint(0, 10, (2, 20))
data['ntype2', 'etype2', 'ntype2'].edge_index = torch.randint(0, 10, (2, 20))
data['ntype3', 'etype3', 'ntype3'].edge_weight = torch.rand(20)
data['ntype3', 'etype3', 'ntype3'].edge_weight = torch.rand(20)

out_dict = conv(data.x_dict, data.edge_index_dict, data.edge_weight_dict)

**Reply by amalislam675** [reply]
@rusty1s , I am using PYG 2.3.1. this example works fine, which you provided. May be there is an issue with my data when I create edge weights. I will again recheck my data.

**Reply by amalislam675** [reply]
@rusty1s , however, your shared example works fine on my end when I try to run GCNConv or GraphConv from PYG. But, when I run this model on my data, I found the same error for GCNConv which I shared above. However, If I run GraphConv from PYG on my data, I find this error, "TypeError: can't convert cuda:3 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.". However, in my code I already did this this using score.cpu().detach().numpy(). Can you help me to fix this error?
 Detail error message is attached below: 
`Traceback (most recent call last):
  File "/pytorchgeometric/example/examples/example_code/src/method/main.py", line 243, in <module>
    train(model, optimizer, train_data)
  File "/pytorchgeometric/example/examples/example_code/src/method/main.py", line 61, in train
    trn_loss = train_epoch()
  File "/pytorchgeometric/example/examples/example_code/src/method/main.py", line 41, in train_epoch
    score = model(data.x_dict, data.edge_index_dict, data.edge_weight_dict)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Project1/pytorchgeometric/example/examples/example_code/src/method/model2.py", line 20, in forward
    h = self.conv1(x_dict, edge_index_dict, edge_weight_dict)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch_geometric/nn/conv/hetero_conv.py", line 136, in forward
    out = conv(x_dict[src], edge_index, *args, **kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch_geometric/nn/conv/graph_conv.py", line 86, in forward
    out = self.propagate(edge_index, x=x, edge_weight=edge_weight,
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch_geometric/nn/conv/message_passing.py", line 467, in propagate
    out = self.message(**msg_kwargs)
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch_geometric/nn/conv/graph_conv.py", line 97, in message
    return x_j if edge_weight is None else edge_weight.view(-1, 1) * x_j
  File "/.conda/envs/foo/lib/python3.10/site-packages/torch/_tensor.py", line 757, in __array__
    return self.numpy()
TypeError: can't convert cuda:3 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
`

In start, three types of graphs are in the form of  weighted adjacency matrix, and I converted them to edge_index and edge_weight format using this below mentioned code:
`graph1_1 = torch.FloatTensor(graph1)
    edge_index1 = graph1_1.nonzero().t()
    edge_index1 = torch.LongTensor(edge_index1)
    row, col = edge_index1
    edge_weight1 = graph1_1[row, col]
    edge_weight1 = torch.FloatTensor(edge_weight1)
    
    
    graph1_2 = torch.FloatTensor(graph2)
    edge_index2 = graph1_2.nonzero().t()
    edge_index2 = torch.LongTensor(edge_index2)
    row, col = edge_index2
    edge_weight2= graph1_2[row, col]
    edge_weight2 = torch.FloatTensor(edge_weight2)
    
    
    graph1_3 = torch.FloatTensor(graph3)
    edge_index3 = graph1_3.nonzero().t()
    edge_index3 = torch.LongTensor(edge_index3)
    row, col = edge_index3
    edge_weight3= graph1_3[row, col]
    edge_weight3 = torch.FloatTensor(edge_weight3)` 

My HeteroData Object is of this form
`HeteroData(
  ntype1={ x=[4705, 4705] },
  ntype2={ x=[4705, 4705] },
  ntype3={ x=[4705, 4705] },
  (ntype1, etype1, ntype1)={
    edge_index=[2, 1511590],
    edge_weight=[1511590]
  },
  (ntype2, etype2, ntype2)={
    edge_index=[2, 1645254],
    edge_weight=[1645254]
  },
  (ntype3, etype3, ntype3)={
    edge_index=[2, 322990],
    edge_weight=[322990]
  }
)`

My model code is shown below:
`import torch
from torch_geometric.nn import HeteroConv, GraphConv
class Model(torch.nn.Module):
  def __init__(self, out_channels, hidden_channels=512):
    super().__init__()
    self.conv1  = HeteroConv({
      ('ntype1', 'etype1', 'ntype1'): GraphConv(-1, hidden_channels),
      ('ntype2', 'etype2', 'ntype2'): GraphConv(-1, hidden_channels),
      ('ntype3', 'etype3', 'ntype3'): GraphConv(-1, hidden_channels)
    }, aggr = 'sum')
    self.conv2 = HeteroConv({
      ('ntype1', 'etype1', 'ntype1'): GraphConv(hidden_channels, hidden_channels),
      ('ntype2', 'etype2', 'ntype2'): GraphConv(hidden_channels, hidden_channels),
      ('ntype3', 'etype3', 'ntype3'): GraphConv(hidden_channels, hidden_channels)
      }, aggr=sum)
    self.lin1 = torch.nn.Linear(hidden_channels*3, out_channels)
    
  def forward(self, x_dict, edge_index_dict, edge_weight_dict):
    h = self.conv1(x_dict, edge_index_dict, edge_weight_dict)
    h = self.conv2(h, edge_index_dict, edge_weight_dict)
  
    feature1 = h['ntype1']
    feature2 = h['ntype2']
    feature3 = h['ntype3']
    
    out = torch.cat((feature1, feature2, feature3), dim=1)
    pred = self.lin1(out)
    return pred `





To transfer the data object from CPU to GPU, I used this below code:
`# Transfer data object to GPU.
    
    data = data.pin_memory()
    data = data.to('cuda:3', non_blocking=True)`

My model object and train epoch function is:

 `model = Model(500).cuda(3)`
  
   More description:
I am using pytorch 1.12.0, numpy 1.25.2, pytorch geometric version 2.3.1 and 3.10 version

**Reply by rusty1s** [reply]
That's weird. Can you ensure that all inputs are indeed PyTorch tensors? Is there a way for you to debug why it complains about `numpy` in the GNN layer?

**Reply by amalislam675** [reply]
@rusty1s , I checked all input types, all inputs type is of pytorch tensor. I donot understand why this error come?

**Reply by rusty1s** [reply]
Do you have a small example to reproduce? I can try to find the issue :)

**Reply by amalislam675** [reply]
@rusty1s , I run this code on my data then error occurs, If I run it on random data then it works fine. In the previous comment, I shared, how I convert my adjacency matrices to PYG edge_index, edge_weight format. I will try to share the example.

**Reply by rusty1s** [reply]
Is it possible for you to upload a subset of your data? Alternatively, you can just share the `HeteroData` object with me via `torch.save`.

**Reply by amalislam675** [reply]
@rusty1s , I will try to share the data.

**Comment by vTechSavyy** [comment]
In case it helps, I was also facing the same error 
`packages/torch_geometric/nn/conv/gcn_conv.py", line 211, in forward edge_index, edge_weight, x.size(self.node_dim), TypeError: 'int' object is not callable` 

In my case, when I was creating the hetero-data I was mistakenly setting the node feature tensors `x` as `numpy` arrays instead of `torch` tensors.
The `HeteroData` class seemed to accept that, but then in the call at `gcn_conv.py", line 211`, the `size()` of `torch::Tensor` is called. 
And in `numpy` arrays the member `size` is an `int` indicating the total number of elements in the array. Hence, the error `'int' object is not callable`. 
For my case the fix was simple, I just had to change the `numpy` arrays to `torch` Tensors :-) (`torch.from_numpy` did the job)
